// electron-forge/dist/makers/win32/squirrel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isSupportedOnCurrentPlatform = undefined;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _bluebird = require('bluebird');

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _ensureOutput = require('../../util/ensure-output');

var _configFn = require('../../util/config-fn');

var _configFn2 = _interopRequireDefault(_configFn);

var _isInstalled = require('../../util/is-installed');

var _isInstalled2 = _interopRequireDefault(_isInstalled);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const isSupportedOnCurrentPlatform = exports.isSupportedOnCurrentPlatform = (() => {
  var _ref = (0, _bluebird.coroutine)(function* () {
    return (0, _isInstalled2.default)('electron-winstaller');
  });

  return function isSupportedOnCurrentPlatform() {
    return _ref.apply(this, arguments);
  };
})();

exports.default = (() => {
  var _ref2 = (0, _bluebird.coroutine)(function* ({ dir, appName, targetArch, forgeConfig, packageJSON }) {
    var _require = require('electron-winstaller');

    const createWindowsInstaller = _require.createWindowsInstaller;


    const outPath = _path2.default.resolve(dir, `../make/squirrel.windows/${targetArch}`);
    yield (0, _ensureOutput.ensureDirectory)(outPath);

    const winstallerConfig = (0, _assign2.default)({
      name: appName,
      noMsi: true,
      exe: `${appName}.exe`,
      setupExe: `${appName}-${packageJSON.version} Setup.exe`
    }, (0, _configFn2.default)(forgeConfig.electronWinstallerConfig, targetArch), {
      appDirectory: dir,
      outputDirectory: outPath
    });

    yield createWindowsInstaller(winstallerConfig);

    const artifacts = [_path2.default.resolve(outPath, 'RELEASES'), _path2.default.resolve(outPath, winstallerConfig.setupExe || `${appName}Setup.exe`), _path2.default.resolve(outPath, `${winstallerConfig.name}-${packageJSON.version}-full.nupkg`)];
    const deltaPath = _path2.default.resolve(outPath, `${winstallerConfig.name}-${packageJSON.version}-delta.nupkg`);
    if (winstallerConfig.remoteReleases || (yield _fsExtra2.default.pathExists(deltaPath))) {
      artifacts.push(deltaPath);
    }
    const msiPath = _path2.default.resolve(outPath, winstallerConfig.setupMsi || `${appName}Setup.msi`);
    if (!winstallerConfig.noMsi && (yield _fsExtra2.default.pathExists(msiPath))) {
      artifacts.push(msiPath);
    }
    return artifacts;
  });

  return function (_x) {
    return _ref2.apply(this, arguments);
  };
})();
