// electron-forge/dist/makers/win32/wix.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isSupportedOnCurrentPlatform = undefined;

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _bluebird = require('bluebird');

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _ensureOutput = require('../../util/ensure-output');

var _authorName = require('../../util/author-name');

var _authorName2 = _interopRequireDefault(_authorName);

var _configFn = require('../../util/config-fn');

var _configFn2 = _interopRequireDefault(_configFn);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// electron-wix-msi doesn't set its 'os' field even though it only runs on win32
const isSupportedOnCurrentPlatform = exports.isSupportedOnCurrentPlatform = (() => {
  var _ref = (0, _bluebird.coroutine)(function* () {
    return process.platform === 'win32';
  });

  return function isSupportedOnCurrentPlatform() {
    return _ref.apply(this, arguments);
  };
})();

exports.default = (() => {
  var _ref2 = (0, _bluebird.coroutine)(function* ({ dir, appName, targetArch, forgeConfig, packageJSON }) {
    var _require = require('electron-wix-msi');

    const MSICreator = _require.MSICreator;


    const outPath = _path2.default.resolve(dir, `../make/wix/${targetArch}`);
    yield (0, _ensureOutput.ensureDirectory)(outPath);

    const creator = new MSICreator((0, _assign2.default)({
      description: packageJSON.description,
      name: appName,
      version: packageJSON.version,
      manufacturer: (0, _authorName2.default)(packageJSON.author),
      exe: `${appName}.exe`
    }, (0, _configFn2.default)(forgeConfig.electronWixMSIConfig, targetArch), {
      appDirectory: dir,
      outputDirectory: outPath
    }));

    yield creator.create();

    var _ref3 = yield creator.compile();

    const msiFile = _ref3.msiFile;


    return [msiFile];
  });

  return function (_x) {
    return _ref2.apply(this, arguments);
  };
})();
