// electron-forge/dist/util/check-system.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _bluebird = require('bluebird');

let checkGitExists = (() => {
  var _ref = (0, _bluebird.coroutine)(function* () {
    return new _promise2.default(function (resolve) {
      (0, _child_process.exec)('git --version', function (err) {
        if (err) return resolve(false);
        resolve(true);
      });
    });
  });

  return function checkGitExists() {
    return _ref.apply(this, arguments);
  };
})();

let checkNodeVersion = (() => {
  var _ref2 = (0, _bluebird.coroutine)(function* () {
    return _promise2.default.resolve(_semver2.default.gt(process.versions.node, '6.0.0'));
  });

  return function checkNodeVersion() {
    return _ref2.apply(this, arguments);
  };
})();

let checkPackageManagerVersion = (() => {
  var _ref3 = (0, _bluebird.coroutine)(function* (ora) {
    return (0, _yarnOrNpm.yarnOrNpmSpawn)(['--version']).then(function (version) {
      if ((0, _yarnOrNpm.hasYarn)()) {
        warnIfPackageManagerIsntAKnownGoodVersion('Yarn', version, YARN_WHITELISTED_VERSIONS, ora);
      } else {
        warnIfPackageManagerIsntAKnownGoodVersion('NPM', version, NPM_WHITELISTED_VERSIONS, ora);
      }

      return true;
    });
  });

  return function checkPackageManagerVersion(_x) {
    return _ref3.apply(this, arguments);
  };
})();

exports.validPackageManagerVersion = validPackageManagerVersion;

var _child_process = require('child_process');

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _semver = require('semver');

var _semver2 = _interopRequireDefault(_semver);

var _yarnOrNpm = require('./yarn-or-npm');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const d = (0, _debug2.default)('electron-forge:check-system');

const NPM_WHITELISTED_VERSIONS = {
  all: '^3.0.0 || ^4.0.0 || ~5.1.0 || ~5.2.0 || >= 5.4.2',
  darwin: '>= 5.4.0',
  linux: '>= 5.4.0'
};
const YARN_WHITELISTED_VERSIONS = {
  all: '0.23.3 || 0.24.6 || >= 1.0.0',
  darwin: '0.27.5',
  linux: '0.27.5'
};

function validPackageManagerVersion(packageManager, version, whitelistedVersions, ora) {
  try {
    return _semver2.default.satisfies(version, whitelistedVersions);
  } catch (e) {
    ora.warn(`Could not check ${packageManager} version "${version}", assuming incompatible`);
    d(`Exception while checking version: ${e}`);
    return false;
  }
}

function warnIfPackageManagerIsntAKnownGoodVersion(packageManager, version, whitelistedVersions, ora) {
  const osVersions = whitelistedVersions[process.platform];
  const versions = osVersions ? `${whitelistedVersions.all} || ${osVersions}` : whitelistedVersions.all;
  const versionString = version.toString();
  if (!validPackageManagerVersion(packageManager, versionString, versions, ora)) {
    ora.warn(`You are using ${packageManager}, but not a known good version.\n` + `The known versions that work with Electron Forge are: ${versions}`);
  }
}

exports.default = (() => {
  var _ref4 = (0, _bluebird.coroutine)(function* (ora) {
    return (yield _promise2.default.all([checkGitExists(ora), checkNodeVersion(ora), checkPackageManagerVersion(ora)])).every(function (check) {
      return check;
    });
  });

  return function (_x2) {
    return _ref4.apply(this, arguments);
  };
})();
