// electron-forge/dist/util/ora.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fakeOra = undefined;

require('colors');

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _logSymbols = require('log-symbols');

var _logSymbols2 = _interopRequireDefault(_logSymbols);

var _ora = require('ora');

var _ora2 = _interopRequireDefault(_ora);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const d = (0, _debug2.default)('electron-forge:lifecycle');

const useFakeOra = process.env.DEBUG && process.env.DEBUG.includes('electron-forge');

if (useFakeOra) {
  console.warn('WARNING: DEBUG environment variable detected.  Progress indicators will be sent over electron-forge:lifecycle'.red);
}

const fakeOra = exports.fakeOra = name => {
  const fake = {
    start: () => {
      d('Process Started:', name);
      return fake;
    },
    warn: msg => {
      console.warn(_logSymbols2.default.warning, msg.yellow);
    },
    fail: () => {
      d(`Process Failed: ${name}`.red);
      return fake;
    },
    succeed: () => {
      d('Process Succeeded:', name);
      return fake;
    },
    stop: () => {
      d('Process Stopped:', name);
      return fake;
    }
  };
  return fake;
};

exports.default = useFakeOra ? fakeOra : _ora2.default;
