// electron-forge/dist/util/publish-state.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _bluebird = require('bluebird');

var _crypto = require('crypto');

var _crypto2 = _interopRequireDefault(_crypto);

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const EXTENSION = '.forge.publish';

class PublishState {
  static loadFromDirectory(directory, rootDir) {
    return (0, _bluebird.coroutine)(function* () {
      if (!(yield _fsExtra2.default.exists(directory))) {
        throw new Error(`Attempted to load publish state from a missing directory: ${directory}`);
      }

      const publishes = [];
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = (0, _getIterator3.default)((yield _fsExtra2.default.readdir(directory))), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          const dirName = _step.value;

          const subDir = _path2.default.resolve(directory, dirName);
          const states = [];
          if ((yield _fsExtra2.default.stat(subDir)).isDirectory()) {
            const filePaths = (yield _fsExtra2.default.readdir(subDir)).filter(function (fileName) {
              return fileName.endsWith(EXTENSION);
            }).map(function (fileName) {
              return _path2.default.resolve(subDir, fileName);
            });

            var _iteratorNormalCompletion2 = true;
            var _didIteratorError2 = false;
            var _iteratorError2 = undefined;

            try {
              for (var _iterator2 = (0, _getIterator3.default)(filePaths), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
                const filePath = _step2.value;

                const state = new PublishState(filePath);
                yield state.load();
                state.state.artifacts = state.state.artifacts.map(function (artifactPath) {
                  return _path2.default.resolve(rootDir, artifactPath);
                });
                states.push(state);
              }
            } catch (err) {
              _didIteratorError2 = true;
              _iteratorError2 = err;
            } finally {
              try {
                if (!_iteratorNormalCompletion2 && _iterator2.return) {
                  _iterator2.return();
                }
              } finally {
                if (_didIteratorError2) {
                  throw _iteratorError2;
                }
              }
            }
          }
          publishes.push(states);
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      return publishes;
    })();
  }

  static saveToDirectory(directory, artifacts, rootDir) {
    return (0, _bluebird.coroutine)(function* () {
      const id = _crypto2.default.createHash('SHA256').update((0, _stringify2.default)(artifacts)).digest('hex');
      var _iteratorNormalCompletion3 = true;
      var _didIteratorError3 = false;
      var _iteratorError3 = undefined;

      try {
        for (var _iterator3 = (0, _getIterator3.default)(artifacts), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
          const artifact = _step3.value;

          artifact.artifacts = artifact.artifacts.map(function (artifactPath) {
            return _path2.default.relative(rootDir, artifactPath);
          });
          const state = new PublishState(_path2.default.resolve(directory, id, 'null'), '', false);
          state.setState(artifact);
          yield state.saveToDisk();
        }
      } catch (err) {
        _didIteratorError3 = true;
        _iteratorError3 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion3 && _iterator3.return) {
            _iterator3.return();
          }
        } finally {
          if (_didIteratorError3) {
            throw _iteratorError3;
          }
        }
      }
    })();
  }

  constructor(filePath, hasHash = true) {
    this.dir = _path2.default.dirname(filePath);
    this.path = filePath;
    this.hasHash = hasHash;
  }

  generateHash() {
    const content = (0, _stringify2.default)(this.state || {});
    return _crypto2.default.createHash('SHA256').update(content).digest('hex');
  }

  setState(state) {
    this.state = state;
  }

  load() {
    var _this = this;

    return (0, _bluebird.coroutine)(function* () {
      _this.state = yield _fsExtra2.default.readJson(_this.path);
    })();
  }

  saveToDisk() {
    var _this2 = this;

    return (0, _bluebird.coroutine)(function* () {
      if (!_this2.hasHash) {
        _this2.path = _path2.default.resolve(_this2.dir, `${_this2.generateHash()}${EXTENSION}`);
        _this2.hasHash = true;
      }

      yield _fsExtra2.default.mkdirs(_path2.default.dirname(_this2.path));
      yield _fsExtra2.default.writeJson(_this2.path, _this2.state);
    })();
  }
}
exports.default = PublishState;
