// electron-forge/dist/publishers/github.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _bluebird = require('bluebird');

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _mimeTypes = require('mime-types');

var _mimeTypes2 = _interopRequireDefault(_mimeTypes);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _oraHandler = require('../util/ora-handler');

var _oraHandler2 = _interopRequireDefault(_oraHandler);

var _github = require('../util/github');

var _github2 = _interopRequireDefault(_github);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = (() => {
  var _ref = (0, _bluebird.coroutine)(function* ({ artifacts, packageJSON, forgeConfig, authToken, tag }) {
    if (!(forgeConfig.github_repository && typeof forgeConfig.github_repository === 'object' && forgeConfig.github_repository.owner && forgeConfig.github_repository.name)) {
      throw 'In order to publish to github you must set the "github_repository.owner" and "github_repository.name" properties in your forge config. See the docs for more info'; // eslint-disable-line
    }

    const github = new _github2.default(authToken, true, forgeConfig.github_repository.options);

    let release;
    yield (0, _oraHandler2.default)('Searching for target release', (0, _bluebird.coroutine)(function* () {
      try {
        release = (yield github.getGitHub().repos.getReleases({
          owner: forgeConfig.github_repository.owner,
          repo: forgeConfig.github_repository.name,
          per_page: 100
        })).data.find(function (testRelease) {
          return testRelease.tag_name === (tag || `v${packageJSON.version}`);
        });
        if (!release) {
          throw { code: 404 };
        }
      } catch (err) {
        if (err.code === 404) {
          // Release does not exist, let's make it
          release = (yield github.getGitHub().repos.createRelease({
            owner: forgeConfig.github_repository.owner,
            repo: forgeConfig.github_repository.name,
            tag_name: tag || `v${packageJSON.version}`,
            name: tag || `v${packageJSON.version}`,
            draft: forgeConfig.github_repository.draft !== false,
            prerelease: forgeConfig.github_repository.prerelease === true
          })).data;
        } else {
          // Unknown error
          throw err;
        }
      }
    }));

    let uploaded = 0;
    yield (0, _oraHandler2.default)(`Uploading Artifacts ${uploaded}/${artifacts.length}`, (() => {
      var _ref3 = (0, _bluebird.coroutine)(function* (uploadSpinner) {
        const updateSpinner = function updateSpinner() {
          uploadSpinner.text = `Uploading Artifacts ${uploaded}/${artifacts.length}`; // eslint-disable-line
        };

        yield _promise2.default.all(artifacts.map(function (artifactPath) {
          return new _promise2.default((() => {
            var _ref4 = (0, _bluebird.coroutine)(function* (resolve) {
              const done = function done() {
                uploaded += 1;
                updateSpinner();
                resolve();
              };
              if (release.assets.find(function (asset) {
                return asset.name === _path2.default.basename(artifactPath);
              })) {
                return done();
              }
              yield github.getGitHub().repos.uploadAsset({
                url: release.upload_url,
                file: _fsExtra2.default.createReadStream(artifactPath),
                contentType: _mimeTypes2.default.lookup(artifactPath) || 'application/octet-stream',
                contentLength: (yield _fsExtra2.default.stat(artifactPath)).size,
                name: _path2.default.basename(artifactPath)
              });
              return done();
            });

            return function (_x3) {
              return _ref4.apply(this, arguments);
            };
          })());
        }));
      });

      return function (_x2) {
        return _ref3.apply(this, arguments);
      };
    })());
  });

  return function (_x) {
    return _ref.apply(this, arguments);
  };
})();
