// electron-forge/dist/api/import.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _bluebird = require('bluebird');

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _inquirer = require('inquirer');

var _inquirer2 = _interopRequireDefault(_inquirer);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _initGit = require('../init/init-git');

var _initGit2 = _interopRequireDefault(_initGit);

var _initNpm = require('../init/init-npm');

var _forgeConfig = require('../util/forge-config');

var _oraHandler = require('../util/ora-handler');

var _oraHandler2 = _interopRequireDefault(_oraHandler);

var _messages = require('../util/messages');

var _installDependencies = require('../util/install-dependencies');

var _installDependencies2 = _interopRequireDefault(_installDependencies);

var _readPackageJson = require('../util/read-package-json');

var _readPackageJson2 = _interopRequireDefault(_readPackageJson);

var _confirmIfInteractive = require('../util/confirm-if-interactive');

var _confirmIfInteractive2 = _interopRequireDefault(_confirmIfInteractive);

var _yarnOrNpm = require('../util/yarn-or-npm');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const d = (0, _debug2.default)('electron-forge:import');

/**
 * @typedef {Object} ImportOptions
 * @property {string} [dir=process.cwd()] The path to the app to be imported
 * @property {boolean} [interactive=false] Whether to use sensible defaults or prompt the user visually
 * @property {boolean} [updateScripts=true] Whether to update the modules package.json scripts to be electron-forge commands
 * @property {string} [outDir=`${dir}/out`] The path to the directory containing generated distributables
 */

/**
 * Attempt to import a given module directory to the Electron Forge standard.
 *
 * - Replaces the prebuilt electron package with the one that integrates with `electron-compile`
 * - Sets up `git` and the correct NPM dependencies
 * - Adds a template forge config to `package.json`
 *
 * @param {ImportOptions} providedOptions - Options for the import method
 * @return {Promise} Will resolve when the import process is complete
 */

exports.default = (() => {
  var _ref = (0, _bluebird.coroutine)(function* (providedOptions = {}) {
    var _Object$assign = (0, _assign2.default)({
      dir: process.cwd(),
      interactive: false,
      updateScripts: true
    }, providedOptions);

    const dir = _Object$assign.dir,
          interactive = _Object$assign.interactive,
          updateScripts = _Object$assign.updateScripts;


    const outDir = providedOptions.outDir || 'out';
    _oraHandler2.default.interactive = interactive;

    d(`Attempting to import project in: ${dir}`);
    if (!(yield _fsExtra2.default.pathExists(dir)) || !(yield _fsExtra2.default.pathExists(_path2.default.resolve(dir, 'package.json')))) {
      throw `We couldn't find a project in: ${dir}`;
    }

    // eslint-disable-next-line max-len
    const confirm = yield (0, _confirmIfInteractive2.default)(interactive, `WARNING: We will now attempt to import: "${dir}".  This will involve modifying some files, are you sure you want to continue?`);

    if (!confirm) {
      process.exit(0);
    }

    yield (0, _initGit2.default)(dir);

    let packageJSON = yield (0, _readPackageJson2.default)(dir);
    if (packageJSON.config && packageJSON.config.forge) {
      (0, _messages.warn)(interactive, 'It looks like this project is already configured for "electron-forge"'.green);
      const shouldContinue = yield (0, _confirmIfInteractive2.default)(interactive, 'Are you sure you want to continue?');

      if (!shouldContinue) {
        process.exit(0);
      }
    }

    // eslint-disable-next-line max-len
    const shouldChangeMain = yield (0, _confirmIfInteractive2.default)(interactive, 'Do you want us to change the "main" attribute of your package.json?  If you are currently using babel and pointing to a "build" directory say yes.', false);
    if (shouldChangeMain) {
      var _ref2 = yield _inquirer2.default.createPromptModule()({
        type: 'input',
        name: 'newMain',
        default: packageJSON.main,
        message: 'Enter the relative path to your uncompiled main file'
      });

      const newMain = _ref2.newMain;

      packageJSON.main = newMain;
    }

    packageJSON.dependencies = packageJSON.dependencies || {};
    packageJSON.devDependencies = packageJSON.devDependencies || {};

    const keys = (0, _keys2.default)(packageJSON.dependencies).concat((0, _keys2.default)(packageJSON.devDependencies));
    const buildToolPackages = {
      'electron-builder': 'provides mostly equivalent functionality',
      'electron-download': 'already uses this module as a transitive dependency',
      'electron-installer-debian': 'already uses this module as a transitive dependency',
      'electron-installer-dmg': 'already uses this module as a transitive dependency',
      'electron-installer-flatpak': 'already uses this module as a transitive dependency',
      'electron-installer-redhat': 'already uses this module as a transitive dependency',
      'electron-osx-sign': 'already uses this module as a transitive dependency',
      'electron-packager': 'already uses this module as a transitive dependency',
      'electron-winstaller': 'already uses this module as a transitive dependency'
    };

    let electronName;
    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = (0, _getIterator3.default)(keys), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        const key = _step.value;

        if (key === 'electron' || key === 'electron-prebuilt') {
          delete packageJSON.dependencies[key];
          delete packageJSON.devDependencies[key];
          electronName = key;
        } else if (buildToolPackages[key]) {
          const explanation = buildToolPackages[key];
          // eslint-disable-next-line max-len
          const shouldRemoveDependency = yield (0, _confirmIfInteractive2.default)(interactive, `Do you want us to remove the "${key}" dependency in package.json? Electron Forge ${explanation}.`);

          if (shouldRemoveDependency) {
            delete packageJSON.dependencies[key];
            delete packageJSON.devDependencies[key];
          }
        }
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator.return) {
          _iterator.return();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }

    packageJSON.scripts = packageJSON.scripts || {};
    d('reading current scripts object:', packageJSON.scripts);

    const updatePackageScript = (() => {
      var _ref3 = (0, _bluebird.coroutine)(function* (scriptName, newValue) {
        if (packageJSON.scripts[scriptName] !== newValue) {
          // eslint-disable-next-line max-len
          const shouldUpdate = yield (0, _confirmIfInteractive2.default)(interactive, `Do you want us to update the "${scriptName}" script to instead call the electron-forge task "${newValue}"`, updateScripts);
          if (shouldUpdate) {
            packageJSON.scripts[scriptName] = newValue;
          }
        }
      });

      return function updatePackageScript(_x, _x2) {
        return _ref3.apply(this, arguments);
      };
    })();

    yield updatePackageScript('start', 'electron-forge start');
    yield updatePackageScript('package', 'electron-forge package');
    yield updatePackageScript('make', 'electron-forge make');

    d('forgified scripts object:', packageJSON.scripts);

    const writeChanges = (() => {
      var _ref4 = (0, _bluebird.coroutine)(function* () {
        yield (0, _oraHandler2.default)('Writing modified package.json file', (0, _bluebird.coroutine)(function* () {
          yield _fsExtra2.default.writeJson(_path2.default.resolve(dir, 'package.json'), packageJSON, { spaces: 2 });
        }));
      });

      return function writeChanges() {
        return _ref4.apply(this, arguments);
      };
    })();

    let electronVersion;
    if (electronName) {
      const electronPackageJSON = yield (0, _readPackageJson2.default)(_path2.default.resolve(dir, 'node_modules', electronName));
      electronVersion = electronPackageJSON.version;
      packageJSON.devDependencies['electron-prebuilt-compile'] = electronVersion;
    }

    yield writeChanges();

    if (electronName) {
      yield (0, _oraHandler2.default)('Pruning deleted modules', (0, _bluebird.coroutine)(function* () {
        d('attempting to prune node_modules in:', dir);
        yield (0, _yarnOrNpm.yarnOrNpmSpawn)((0, _yarnOrNpm.hasYarn)() ? [] : ['prune'], {
          cwd: dir,
          stdio: 'ignore'
        });
      }));
    }

    yield (0, _oraHandler2.default)('Installing dependencies', (0, _bluebird.coroutine)(function* () {
      d('deleting old dependencies forcefully');
      yield _fsExtra2.default.remove(_path2.default.resolve(dir, 'node_modules/.bin/electron'));
      yield _fsExtra2.default.remove(_path2.default.resolve(dir, 'node_modules/.bin/electron.cmd'));

      if (electronName) {
        yield _fsExtra2.default.remove(_path2.default.resolve(dir, 'node_modules', electronName));
      }

      d('installing dependencies');
      yield (0, _installDependencies2.default)(dir, _initNpm.deps);

      d('installing devDependencies');
      yield (0, _installDependencies2.default)(dir, _initNpm.devDeps, true);

      d('installing exactDevDependencies');
      yield (0, _installDependencies2.default)(dir, _initNpm.exactDevDeps.map(function (dep) {
        if (dep === 'electron-prebuilt-compile') {
          return `${dep}@${electronVersion || 'latest'}`;
        }

        return dep;
      }), true, true);
    }));

    packageJSON = yield (0, _readPackageJson2.default)(dir);

    if (!packageJSON.version) {
      (0, _messages.warn)(interactive, "Please set the 'version' in your application's package.json".yellow);
    }

    packageJSON.config = packageJSON.config || {};
    const templatePackageJSON = yield (0, _readPackageJson2.default)(_path2.default.resolve(__dirname, '../../tmpl'));
    packageJSON.config.forge = templatePackageJSON.config.forge;
    (0, _forgeConfig.setInitialForgeConfig)(packageJSON);

    yield writeChanges();

    yield (0, _oraHandler2.default)('Fixing .gitignore', (0, _bluebird.coroutine)(function* () {
      if (yield _fsExtra2.default.pathExists(_path2.default.resolve(dir, '.gitignore'))) {
        const gitignore = yield _fsExtra2.default.readFile(_path2.default.resolve(dir, '.gitignore'));
        if (!gitignore.includes(outDir)) {
          yield _fsExtra2.default.writeFile(_path2.default.resolve(dir, '.gitignore'), `${gitignore}\n${outDir}/`);
        }
      }
    }));

    let babelConfig = packageJSON.babel;
    const babelPath = _path2.default.resolve(dir, '.babelrc');
    if (!babelConfig && (yield _fsExtra2.default.pathExists(babelPath))) {
      babelConfig = yield _fsExtra2.default.readJson(babelPath, 'utf8');
    }

    if (babelConfig) {
      yield (0, _oraHandler2.default)('Porting original babel config', (0, _bluebird.coroutine)(function* () {
        let compileConfig = {};
        const compilePath = _path2.default.resolve(dir, '.compilerc');
        if (yield _fsExtra2.default.pathExists(compilePath)) {
          compileConfig = yield _fsExtra2.default.readJson(compilePath, 'utf8');
        }

        yield _fsExtra2.default.writeJson(compilePath, (0, _assign2.default)(compileConfig, {
          'application/javascript': babelConfig
        }), { spaces: 2 });
      }));

      (0, _messages.info)(interactive, 'NOTE: You might be able to remove your `.compilerc` file completely if you are only using the `es2016` and `react` presets'.yellow);
    }

    (0, _messages.info)(interactive, `

We have ATTEMPTED to convert your app to be in a format that electron-forge understands.
Nothing much will have changed but we added the ${'"electron-prebuilt-compile"'.cyan} dependency.  This is \
the dependency you must version bump to get newer versions of Electron.


We also tried to import any build tooling you already had but we can't get everything.  You might need to convert any CLI/gulp/grunt tasks yourself.

Also please note if you are using \`preload\` scripts you need to follow the steps outlined \
at https://github.com/electron-userland/electron-forge/wiki/Using-%27preload%27-scripts

Thanks for using ${'"electron-forge"'.green}!!!`);
  });

  return function () {
    return _ref.apply(this, arguments);
  };
})();
