// electron-forge/dist/api/install.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _bluebird = require('bluebird');

require('colors');

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _nodeFetch = require('node-fetch');

var _nodeFetch2 = _interopRequireDefault(_nodeFetch);

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _inquirer = require('inquirer');

var _inquirer2 = _interopRequireDefault(_inquirer);

var _nugget = require('nugget');

var _nugget2 = _interopRequireDefault(_nugget);

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _pify = require('pify');

var _pify2 = _interopRequireDefault(_pify);

var _semver = require('semver');

var _semver2 = _interopRequireDefault(_semver);

var _oraHandler = require('../util/ora-handler');

var _oraHandler2 = _interopRequireDefault(_oraHandler);

var _messages = require('../util/messages');

var _dmg = require('../installers/darwin/dmg');

var _dmg2 = _interopRequireDefault(_dmg);

var _zip = require('../installers/darwin/zip');

var _zip2 = _interopRequireDefault(_zip);

var _deb = require('../installers/linux/deb');

var _deb2 = _interopRequireDefault(_deb);

var _rpm = require('../installers/linux/rpm');

var _rpm2 = _interopRequireDefault(_rpm);

var _exe = require('../installers/win32/exe');

var _exe2 = _interopRequireDefault(_exe);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const d = (0, _debug2.default)('electron-forge:install');

const GITHUB_API = 'https://api.github.com';

/**
 * @typedef {Object} InstallOptions
 * @property {boolean} [interactive=false] Whether to use sensible defaults or prompt the user visually
 * @property {boolean} [prerelease=false] Whether to install prerelease versions
 * @property {string} repo The GitHub repository to install from, in the format owner/name
 * @property {function} chooseAsset A function that must return the asset to use/install from a provided array of compatible GitHub assets
 */

/**
 * Install an Electron application from GitHub. If you leave interactive as `false`, you MUST provide a `chooseAsset` function.
 *
 * @param {InstallOptions} providedOptions - Options for the install method
 * @return {Promise} Will resolve when the install process is complete
 */

exports.default = (() => {
  var _ref = (0, _bluebird.coroutine)(function* (providedOptions = {}) {
    // eslint-disable-next-line prefer-const, no-unused-vars
    var _Object$assign = (0, _assign2.default)({
      interactive: false,
      prerelease: false
    }, providedOptions);

    let interactive = _Object$assign.interactive,
        prerelease = _Object$assign.prerelease,
        repo = _Object$assign.repo,
        chooseAsset = _Object$assign.chooseAsset;

    _oraHandler2.default.interactive = interactive;

    let latestRelease;
    let possibleAssets = [];

    yield (0, _oraHandler2.default)('Searching for Application', (() => {
      var _ref2 = (0, _bluebird.coroutine)(function* (searchSpinner) {
        if (!repo || repo.indexOf('/') === -1) {
          throw 'Invalid repository name, must be in the format owner/name';
        }

        d('searching for repo:', repo);
        let releases;
        try {
          releases = yield (yield (0, _nodeFetch2.default)(`${GITHUB_API}/repos/${repo}/releases`)).json();
        } catch (err) {
          // Ignore error
        }

        if (!releases || releases.message === 'Not Found' || !Array.isArray(releases)) {
          throw `Failed to find releases for repository "${repo}".  Please check the name and try again.`;
        }

        releases = releases.filter(function (release) {
          return !release.prerelease || prerelease;
        });

        const sortedReleases = releases.sort(function (releaseA, releaseB) {
          let tagA = releaseA.tag_name;
          if (tagA.substr(0, 1) === 'v') tagA = tagA.substr(1);
          let tagB = releaseB.tag_name;
          if (tagB.substr(0, 1) === 'v') tagB = tagB.substr(1);
          return _semver2.default.gt(tagB, tagA) ? 1 : -1;
        });
        latestRelease = sortedReleases[0];

        searchSpinner.text = 'Searching for Releases'; // eslint-disable-line

        const assets = latestRelease.assets;
        if (!assets || !Array.isArray(assets)) {
          throw 'Could not find any assets for the latest release';
        }

        const installTargets = {
          win32: [/\.exe$/],
          darwin: [/OSX.*\.zip$/, /darwin.*\.zip$/, /macOS.*\.zip$/, /mac.*\.zip$/, /\.dmg$/],
          linux: [/\.rpm$/, /\.deb$/]
        };

        possibleAssets = assets.filter(function (asset) {
          const targetSuffixes = installTargets[process.platform];
          var _iteratorNormalCompletion = true;
          var _didIteratorError = false;
          var _iteratorError = undefined;

          try {
            for (var _iterator = (0, _getIterator3.default)(targetSuffixes), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
              const suffix = _step.value;

              if (suffix.test(asset.name)) return true;
            }
          } catch (err) {
            _didIteratorError = true;
            _iteratorError = err;
          } finally {
            try {
              if (!_iteratorNormalCompletion && _iterator.return) {
                _iterator.return();
              }
            } finally {
              if (_didIteratorError) {
                throw _iteratorError;
              }
            }
          }

          return false;
        });

        if (possibleAssets.length === 0) {
          throw `Failed to find any installable assets for target platform: ${`${process.platform}`.cyan}`;
        }
      });

      return function (_x) {
        return _ref2.apply(this, arguments);
      };
    })());

    (0, _messages.info)(interactive, `Found latest release${prerelease ? ' (including prereleases)' : ''}: ${latestRelease.tag_name.cyan}`);

    let targetAsset = possibleAssets[0];
    if (possibleAssets.length > 1) {
      if (chooseAsset) {
        targetAsset = yield _promise2.default.resolve(chooseAsset(possibleAssets));
      } else if (interactive) {
        const choices = [];
        possibleAssets.forEach(function (asset) {
          choices.push({ name: asset.name, value: asset.id });
        });

        var _ref3 = yield _inquirer2.default.createPromptModule()({
          type: 'list',
          name: 'assetID',
          message: 'Multiple potential assets found, please choose one from the list below:'.cyan,
          choices
        });

        const assetID = _ref3.assetID;


        targetAsset = possibleAssets.find(function (asset) {
          return asset.id === assetID;
        });
      } else {
        throw 'expected a chooseAsset function to be provided but it was not';
      }
    }

    const tmpdir = _path2.default.resolve(_os2.default.tmpdir(), 'forge-install');
    const pathSafeRepo = repo.replace(/[/\\]/g, '-');
    const filename = `${pathSafeRepo}-${latestRelease.tag_name}-${targetAsset.name}`;

    const fullFilePath = _path2.default.resolve(tmpdir, filename);
    if (!(yield _fsExtra2.default.pathExists(fullFilePath)) || (yield _fsExtra2.default.stat(fullFilePath)).size !== targetAsset.size) {
      yield _fsExtra2.default.mkdirs(tmpdir);

      const nuggetOpts = {
        target: filename,
        dir: tmpdir,
        resume: true,
        strictSSL: true
      };
      yield (0, _pify2.default)(_nugget2.default)(targetAsset.browser_download_url, nuggetOpts);
    }

    yield (0, _oraHandler2.default)('Installing Application', (() => {
      var _ref4 = (0, _bluebird.coroutine)(function* (installSpinner) {
        const installActions = {
          win32: {
            '.exe': _exe2.default
          },
          darwin: {
            '.zip': _zip2.default,
            '.dmg': _dmg2.default
          },
          linux: {
            '.deb': _deb2.default,
            '.rpm': _rpm2.default
          }
        };

        const suffixFnIdent = (0, _keys2.default)(installActions[process.platform]).find(function (suffix) {
          return targetAsset.name.endsWith(suffix);
        });
        yield installActions[process.platform][suffixFnIdent](fullFilePath, installSpinner);
      });

      return function (_x2) {
        return _ref4.apply(this, arguments);
      };
    })());
  });

  return function () {
    return _ref.apply(this, arguments);
  };
})();
