// electron-forge/dist/api/package.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _bluebird = require('bluebird');

require('colors');

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _glob = require('glob');

var _glob2 = _interopRequireDefault(_glob);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _pify = require('pify');

var _pify2 = _interopRequireDefault(_pify);

var _electronPackager = require('electron-packager');

var _electronPackager2 = _interopRequireDefault(_electronPackager);

var _targets = require('electron-packager/targets');

var _forgeConfig = require('../util/forge-config');

var _forgeConfig2 = _interopRequireDefault(_forgeConfig);

var _getElectronVersion = require('../util/get-electron-version');

var _getElectronVersion2 = _interopRequireDefault(_getElectronVersion);

var _hook = require('../util/hook');

var _hook2 = _interopRequireDefault(_hook);

var _messages = require('../util/messages');

var _ora = require('../util/ora');

var _ora2 = _interopRequireDefault(_ora);

var _compileHook = require('../util/compile-hook');

var _compileHook2 = _interopRequireDefault(_compileHook);

var _readPackageJson = require('../util/read-package-json');

var _readPackageJson2 = _interopRequireDefault(_readPackageJson);

var _rebuild = require('../util/rebuild');

var _rebuild2 = _interopRequireDefault(_rebuild);

var _requireSearch = require('../util/require-search');

var _requireSearch2 = _interopRequireDefault(_requireSearch);

var _resolveDir = require('../util/resolve-dir');

var _resolveDir2 = _interopRequireDefault(_resolveDir);

var _outDir = require('../util/out-dir');

var _outDir2 = _interopRequireDefault(_outDir);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const d = (0, _debug2.default)('electron-forge:packager');

/**
 * @typedef {Object} PackageOptions
 * @property {string} [dir=process.cwd()] The path to the app to package
 * @property {boolean} [interactive=false] Whether to use sensible defaults or prompt the user visually
 * @property {string} [arch=process.arch] The target arch
 * @property {string} [platform=process.platform] The target platform.
 * @property {string} [outDir=`${dir}/out`] The path to the output directory for packaged apps
 */

/**
 * Resolves hooks if they are a path to a file (instead of a `Function`).
 */
function resolveHooks(hooks, dir) {
  if (hooks) {
    return hooks.map(hook => typeof hook === 'string' ? (0, _requireSearch2.default)(dir, [hook]) : hook);
  }

  return [];
}

function sequentialHooks(hooks) {
  return [(() => {
    var _ref = (0, _bluebird.coroutine)(function* (...args) {
      const done = args[args.length - 1];
      const passedArgs = args.splice(0, args.length - 1);
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = (0, _getIterator3.default)(hooks), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          const hook = _step.value;

          yield (0, _pify2.default)(hook)(...passedArgs);
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      done();
    });

    return function () {
      return _ref.apply(this, arguments);
    };
  })()];
}

/**
 * Package an Electron application into an platform dependent format.
 *
 * @param {PackageOptions} providedOptions - Options for the Package method
 * @return {Promise} Will resolve when the package process is complete
 */

exports.default = (() => {
  var _ref2 = (0, _bluebird.coroutine)(function* (providedOptions = {}) {
    // eslint-disable-next-line prefer-const, no-unused-vars
    var _Object$assign = (0, _assign2.default)({
      dir: process.cwd(),
      interactive: false,
      arch: (0, _targets.hostArch)(),
      platform: process.platform
    }, providedOptions);

    let dir = _Object$assign.dir,
        interactive = _Object$assign.interactive,
        arch = _Object$assign.arch,
        platform = _Object$assign.platform;


    const ora = interactive ? _ora2.default : _ora.fakeOra;

    let prepareSpinner = ora(`Preparing to Package Application for arch: ${(arch === 'all' ? 'ia32' : arch).cyan}`).start();
    let prepareCounter = 0;

    dir = yield (0, _resolveDir2.default)(dir);
    if (!dir) {
      throw 'Failed to locate compilable Electron application';
    }

    const packageJSON = yield (0, _readPackageJson2.default)(dir);

    if (_path2.default.dirname(require.resolve(_path2.default.resolve(dir, packageJSON.main))) === dir) {
      console.error(`Entry point: ${packageJSON.main}`.red);
      throw 'The entry point to your application ("packageJSON.main") must be in a subfolder not in the top level directory';
    }

    const forgeConfig = yield (0, _forgeConfig2.default)(dir);
    const outDir = providedOptions.outDir || (0, _outDir2.default)(dir, forgeConfig);
    let packagerSpinner;

    const pruneEnabled = !('prune' in forgeConfig.electronPackagerConfig) || forgeConfig.electronPackagerConfig.prune;

    const rebuildHookFn = (() => {
      var _ref3 = (0, _bluebird.coroutine)(function* (buildPath, electronVersion, pPlatform, pArch, done) {
        yield (0, _rebuild2.default)(buildPath, electronVersion, pPlatform, pArch, forgeConfig.electronRebuildConfig);
        packagerSpinner = ora('Packaging Application').start();
        done();
      });

      return function rebuildHookFn(_x, _x2, _x3, _x4, _x5) {
        return _ref3.apply(this, arguments);
      };
    })();

    const afterCopyHooks = [(() => {
      var _ref4 = (0, _bluebird.coroutine)(function* (buildPath, electronVersion, pPlatform, pArch, done) {
        if (packagerSpinner) {
          packagerSpinner.succeed();
          prepareCounter += 1;
          prepareSpinner = ora(`Preparing to Package Application for arch: ${(prepareCounter === 2 ? 'armv7l' : 'x64').cyan}`).start();
        }
        yield _fsExtra2.default.remove(_path2.default.resolve(buildPath, 'node_modules/electron-compile/test'));
        const bins = yield (0, _pify2.default)(_glob2.default)(_path2.default.join(buildPath, '**/.bin/**/*'));
        var _iteratorNormalCompletion2 = true;
        var _didIteratorError2 = false;
        var _iteratorError2 = undefined;

        try {
          for (var _iterator2 = (0, _getIterator3.default)(bins), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
            const bin = _step2.value;

            yield _fsExtra2.default.remove(bin);
          }
        } catch (err) {
          _didIteratorError2 = true;
          _iteratorError2 = err;
        } finally {
          try {
            if (!_iteratorNormalCompletion2 && _iterator2.return) {
              _iterator2.return();
            }
          } finally {
            if (_didIteratorError2) {
              throw _iteratorError2;
            }
          }
        }

        done();
      });

      return function (_x6, _x7, _x8, _x9, _x10) {
        return _ref4.apply(this, arguments);
      };
    })(), (() => {
      var _ref5 = (0, _bluebird.coroutine)(function* (...args) {
        prepareSpinner.succeed();
        yield (0, _compileHook2.default)(dir, ...args);
      });

      return function () {
        return _ref5.apply(this, arguments);
      };
    })()];

    if (!pruneEnabled) {
      afterCopyHooks.push(rebuildHookFn);
    }

    afterCopyHooks.push((() => {
      var _ref6 = (0, _bluebird.coroutine)(function* (buildPath, electronVersion, pPlatform, pArch, done) {
        const copiedPackageJSON = yield (0, _readPackageJson2.default)(buildPath);
        if (copiedPackageJSON.config && copiedPackageJSON.config.forge) {
          delete copiedPackageJSON.config.forge;
        }
        yield _fsExtra2.default.writeJson(_path2.default.resolve(buildPath, 'package.json'), copiedPackageJSON, { spaces: 2 });
        done();
      });

      return function (_x11, _x12, _x13, _x14, _x15) {
        return _ref6.apply(this, arguments);
      };
    })());

    afterCopyHooks.push(...resolveHooks(forgeConfig.electronPackagerConfig.afterCopy, dir));

    const afterPruneHooks = [];

    if (pruneEnabled) {
      afterPruneHooks.push(rebuildHookFn);
      afterPruneHooks.push(...resolveHooks(forgeConfig.electronPackagerConfig.afterPrune, dir));
    }

    const packageOpts = (0, _assign2.default)({
      asar: false,
      overwrite: true
    }, forgeConfig.electronPackagerConfig, {
      afterCopy: sequentialHooks(afterCopyHooks),
      afterExtract: sequentialHooks(resolveHooks(forgeConfig.electronPackagerConfig.afterExtract, dir)),
      afterPrune: sequentialHooks(afterPruneHooks),
      dir,
      arch,
      platform,
      out: outDir,
      electronVersion: yield (0, _getElectronVersion2.default)(dir)
    });
    packageOpts.quiet = true;
    if (packageOpts.all) {
      throw new Error('electronPackagerConfig.all is not supported by Electron Forge.');
    }
    if (typeof packageOpts.asar === 'object' && packageOpts.asar.unpack) {
      throw new Error('electron-compile does not support asar.unpack yet.  Please use asar.unpackDir');
    }

    if (!packageJSON.version && !packageOpts.appVersion) {
      // eslint-disable-next-line max-len
      (0, _messages.warn)(interactive, "Please set 'version' or 'config.forge.electronPackagerConfig.appVersion' in your application's package.json so auto-updates work properly".yellow);
    }

    yield (0, _hook2.default)(forgeConfig, 'generateAssets');
    yield (0, _hook2.default)(forgeConfig, 'prePackage');

    d('packaging with options', packageOpts);

    yield (0, _electronPackager2.default)(packageOpts);

    yield (0, _hook2.default)(forgeConfig, 'postPackage');

    packagerSpinner.succeed();
  });

  return function () {
    return _ref2.apply(this, arguments);
  };
})();
