// electron-forge/dist/api/publish.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _getIterator2 = require('babel-runtime/core-js/get-iterator');

var _getIterator3 = _interopRequireDefault(_getIterator2);

var _bluebird = require('bluebird');

require('colors');

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _oraHandler = require('../util/ora-handler');

var _oraHandler2 = _interopRequireDefault(_oraHandler);

var _forgeConfig = require('../util/forge-config');

var _forgeConfig2 = _interopRequireDefault(_forgeConfig);

var _readPackageJson = require('../util/read-package-json');

var _readPackageJson2 = _interopRequireDefault(_readPackageJson);

var _requireSearch = require('../util/require-search');

var _requireSearch2 = _interopRequireDefault(_requireSearch);

var _resolveDir = require('../util/resolve-dir');

var _resolveDir2 = _interopRequireDefault(_resolveDir);

var _publishState = require('../util/publish-state');

var _publishState2 = _interopRequireDefault(_publishState);

var _outDir = require('../util/out-dir');

var _outDir2 = _interopRequireDefault(_outDir);

var _make = require('./make');

var _make2 = _interopRequireDefault(_make);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const d = (0, _debug2.default)('electron-forge:publish');

/**
 * @typedef {Object} PublishOptions
 * @property {string} [dir=process.cwd()] The path to the app to be published
 * @property {boolean} [interactive=false] Whether to use sensible defaults or prompt the user visually
 * @property {string} [authToken] An authentication token to use when publishing
 * @property {string} [tag=packageJSON.version] The string to tag this release with
 * @property {Array<string>} [publishTargets=[github]] The publish targets
 * @property {MakeOptions} [makeOptions] Options object to passed through to make()
 * @property {string} [outDir=`${dir}/out`] The path to the directory containing generated distributables
 * @property {boolean} [dryRun=false] Whether to generate dry run meta data but not actually publish
 * @property {boolean} [dryRunResume=false] Whether or not to attempt to resume a previously saved `dryRun` and publish
 * @property {MakeResult} [makeResults=null] Provide results from make so that the publish step doesn't run make itself
 */

/**
 * Publish an Electron application into the given target service.
 *
 * @param {PublishOptions} providedOptions - Options for the Publish method
 * @return {Promise} Will resolve when the publish process is complete
 */
const publish = (() => {
  var _ref = (0, _bluebird.coroutine)(function* (providedOptions = {}) {
    // eslint-disable-next-line prefer-const, no-unused-vars
    var _Object$assign = (0, _assign2.default)({
      dir: process.cwd(),
      interactive: false,
      tag: null,
      makeOptions: {},
      publishTargets: null,
      dryRun: false,
      dryRunResume: false,
      makeResults: null
    }, providedOptions);

    let dir = _Object$assign.dir,
        interactive = _Object$assign.interactive,
        authToken = _Object$assign.authToken,
        tag = _Object$assign.tag,
        publishTargets = _Object$assign.publishTargets,
        makeOptions = _Object$assign.makeOptions,
        dryRun = _Object$assign.dryRun,
        dryRunResume = _Object$assign.dryRunResume,
        makeResults = _Object$assign.makeResults;

    _oraHandler2.default.interactive = interactive;

    if (dryRun && dryRunResume) {
      throw 'Can\'t dry run and resume a dry run at the same time';
    }
    if (dryRunResume && makeResults) {
      throw 'Can\'t resume a dry run and use the provided makeResults at the same time';
    }

    let packageJSON = yield (0, _readPackageJson2.default)(dir);

    const forgeConfig = yield (0, _forgeConfig2.default)(dir);
    const outDir = providedOptions.outDir || (0, _outDir2.default)(dir, forgeConfig);
    const dryRunDir = _path2.default.resolve(outDir, 'publish-dry-run');

    if (dryRunResume) {
      d('attempting to resume from dry run');
      const publishes = yield _publishState2.default.loadFromDirectory(dryRunDir, dir);
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = (0, _getIterator3.default)(publishes), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          const publishStates = _step.value;

          d('publishing for given state set');
          yield publish({
            dir,
            interactive,
            authToken,
            tag,
            publishTargets,
            makeOptions,
            dryRun: false,
            dryRunResume: false,
            makeResults: publishStates.map(function ({ state }) {
              return state;
            })
          });
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      return;
    } else if (!makeResults) {
      d('triggering make');
      makeResults = yield (0, _make2.default)((0, _assign2.default)({
        dir,
        interactive
      }, makeOptions));
    } else {
      // Restore values from dry run
      d('restoring publish settings from dry run');

      var _iteratorNormalCompletion2 = true;
      var _didIteratorError2 = false;
      var _iteratorError2 = undefined;

      try {
        for (var _iterator2 = (0, _getIterator3.default)(makeResults), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
          const makeResult = _step2.value;

          packageJSON = makeResult.packageJSON;
          makeOptions.platform = makeResult.platform;
          makeOptions.arch = makeResult.arch;

          var _iteratorNormalCompletion3 = true;
          var _didIteratorError3 = false;
          var _iteratorError3 = undefined;

          try {
            for (var _iterator3 = (0, _getIterator3.default)(makeResult.artifacts), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
              const makePath = _step3.value;

              if (!(yield _fsExtra2.default.exists(makePath))) {
                throw `Attempted to resume a dry run but an artifact (${makePath}) could not be found`;
              }
            }
          } catch (err) {
            _didIteratorError3 = true;
            _iteratorError3 = err;
          } finally {
            try {
              if (!_iteratorNormalCompletion3 && _iterator3.return) {
                _iterator3.return();
              }
            } finally {
              if (_didIteratorError3) {
                throw _iteratorError3;
              }
            }
          }
        }
      } catch (err) {
        _didIteratorError2 = true;
        _iteratorError2 = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion2 && _iterator2.return) {
            _iterator2.return();
          }
        } finally {
          if (_didIteratorError2) {
            throw _iteratorError2;
          }
        }
      }
    }

    if (dryRun) {
      d('saving results of make in dry run state', makeResults);
      yield _fsExtra2.default.remove(dryRunDir);
      yield _publishState2.default.saveToDirectory(dryRunDir, makeResults, dir);
      return;
    }

    dir = yield (0, _resolveDir2.default)(dir);
    if (!dir) {
      throw 'Failed to locate publishable Electron application';
    }

    const artifacts = makeResults.reduce(function (accum, makeResult) {
      accum.push(...makeResult.artifacts);
      return accum;
    }, []);

    if (publishTargets === null) {
      publishTargets = forgeConfig.publish_targets[makeOptions.platform || process.platform];
    }

    var _iteratorNormalCompletion4 = true;
    var _didIteratorError4 = false;
    var _iteratorError4 = undefined;

    try {
      for (var _iterator4 = (0, _getIterator3.default)(publishTargets), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
        const publishTarget = _step4.value;

        let publisher;
        yield (0, _oraHandler2.default)(`Resolving publish target: ${`${publishTarget}`.cyan}`, (0, _bluebird.coroutine)(function* () {
          // eslint-disable-line no-loop-func
          publisher = (0, _requireSearch2.default)(__dirname, [`../publishers/${publishTarget}.js`, `electron-forge-publisher-${publishTarget}`, publishTarget, _path2.default.resolve(dir, publishTarget), _path2.default.resolve(dir, 'node_modules', publishTarget)]);
          if (!publisher) {
            throw `Could not find a publish target with the name: ${publishTarget}`;
          }
        }));

        yield publisher({
          dir,
          artifacts,
          packageJSON,
          forgeConfig,
          authToken,
          tag,
          platform: makeOptions.platform || process.platform,
          arch: makeOptions.arch || process.arch
        });
      }
    } catch (err) {
      _didIteratorError4 = true;
      _iteratorError4 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion4 && _iterator4.return) {
          _iterator4.return();
        }
      } finally {
        if (_didIteratorError4) {
          throw _iteratorError4;
        }
      }
    }
  });

  return function publish() {
    return _ref.apply(this, arguments);
  };
})();

exports.default = publish;
