// electron-forge/dist/api/start.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _bluebird = require('bluebird');

require('colors');

var _child_process = require('child_process');

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _oraHandler = require('../util/ora-handler');

var _oraHandler2 = _interopRequireDefault(_oraHandler);

var _getElectronVersion = require('../util/get-electron-version');

var _getElectronVersion2 = _interopRequireDefault(_getElectronVersion);

var _readPackageJson = require('../util/read-package-json');

var _readPackageJson2 = _interopRequireDefault(_readPackageJson);

var _rebuild = require('../util/rebuild');

var _rebuild2 = _interopRequireDefault(_rebuild);

var _resolveDir = require('../util/resolve-dir');

var _resolveDir2 = _interopRequireDefault(_resolveDir);

var _forgeConfig = require('../util/forge-config');

var _forgeConfig2 = _interopRequireDefault(_forgeConfig);

var _hook = require('../util/hook');

var _hook2 = _interopRequireDefault(_hook);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @typedef {Object} StartOptions
 * @property {string} [dir=process.cwd()] The path to the electron forge project to run
 * @property {string} [appPath='.'] The path (relative to dir) to the electron app to run relative to the project directory
 * @property {boolean} [interactive=false] Whether to use sensible defaults or prompt the user visually
 * @property {boolean} [enableLogging=false] Enables advanced internal Electron debug calls
 * @property {Array<string>} [args] Arguments to pass through to the launched Electron application
 */

/**
 * Start an Electron application.
 *
 * @param {StartOptions} providedOptions - Options for the Publish method
 * @return {Promise} Will resolve when the application is launched
 */
exports.default = (() => {
  var _ref = (0, _bluebird.coroutine)(function* (providedOptions = {}) {
    // eslint-disable-next-line prefer-const, no-unused-vars
    var _Object$assign = (0, _assign2.default)({
      dir: process.cwd(),
      appPath: '.',
      interactive: false,
      enableLogging: false,
      args: [],
      runAsNode: false,
      inspect: false
    }, providedOptions);

    let dir = _Object$assign.dir,
        interactive = _Object$assign.interactive,
        enableLogging = _Object$assign.enableLogging,
        appPath = _Object$assign.appPath,
        args = _Object$assign.args,
        runAsNode = _Object$assign.runAsNode,
        inspect = _Object$assign.inspect;

    _oraHandler2.default.interactive = interactive;

    yield (0, _oraHandler2.default)('Locating Application', (0, _bluebird.coroutine)(function* () {
      dir = yield (0, _resolveDir2.default)(dir);
      if (!dir) {
        throw 'Failed to locate startable Electron application';
      }
    }));

    const packageJSON = yield (0, _readPackageJson2.default)(dir);

    if (!packageJSON.version) {
      throw `Please set your application's 'version' in '${dir}/package.json'.`;
    }

    const forgeConfig = yield (0, _forgeConfig2.default)(dir);
    const electronVersion = yield (0, _getElectronVersion2.default)(dir);

    yield (0, _rebuild2.default)(dir, electronVersion, process.platform, process.arch, forgeConfig.electronRebuildConfig);

    const spawnOpts = {
      cwd: dir,
      stdio: 'inherit',
      env: (0, _assign2.default)({}, process.env, enableLogging ? {
        ELECTRON_ENABLE_LOGGING: true,
        ELECTRON_ENABLE_STACK_DUMPING: true
      } : {})
    };

    if (runAsNode) {
      spawnOpts.env.ELECTRON_RUN_AS_NODE = true;
    } else {
      delete spawnOpts.env.ELECTRON_RUN_AS_NODE;
    }

    if (inspect) {
      args = ['--inspect'].concat(args);
    }

    let spawned;

    yield (0, _hook2.default)(forgeConfig, 'generateAssets');

    yield (0, _oraHandler2.default)('Launching Application', (0, _bluebird.coroutine)(function* () {
      spawned = (0, _child_process.spawn)(process.execPath, [_path2.default.resolve(dir, 'node_modules/electron-prebuilt-compile/lib/cli'), appPath].concat(args), spawnOpts);
    }));

    return spawned;
  });

  return function () {
    return _ref.apply(this, arguments);
  };
})();
